server/aws-lsp-identity/src/sso/cache/fileSystemSsoCache.ts

```diff
@@ -0,0 +1,107 @@
+
+export class FileSystemSsoCache implements SsoCache {
+    async getSsoClientRegistration(
+        clientName: string,
+        ssoRegion: string,
+        ssoStartUrl: string
+    ): Promise<SsoClientRegistration | undefined> {
+        const id = toSsoClientRegistrationCacheId(clientName, ssoRegion, ssoStartUrl)
+        return await tryGetAsync(async () => {
+            const registration = await getSsoClientRegistrationFromFile(id)
+            return ssoClientRegistrationDuckTyper.eval(registration) ? registration : undefined
+        })
+    }
+
+    async setSsoClientRegistration(
+        clientName: string,
+        ssoRegion: string,
+        ssoStartUrl: string,
+        clientRegistration: SsoClientRegistration
+    ): Promise<void> {
+        if (!clientName || !ssoRegion || !ssoStartUrl || !ssoClientRegistrationDuckTyper.eval(clientRegistration)) {
+            return
+        }
+
+        const id = toSsoClientRegistrationCacheId(clientName, ssoRegion, ssoStartUrl)
+        await tryAsync(
+            () => writeSsoObjectToFile(id, clientRegistration),
+            error => AwsError.wrap(error, AwsErrorCodes.E_CANNOT_WRITE_SSO_CACHE)
+        )
+    }
+
+    async getSsoToken(ssoSessionName: string): Promise<SSOToken | undefined> {
+        return await tryGetAsync(async () => {
+            const ssoToken = await getSSOTokenFromFile(ssoSessionName)
+            return ssoTokenDuckTyper.eval(ssoToken) ? ssoToken : undefined
+        })
+    }
+
+    async setSsoToken(ssoSessionName: string, ssoToken: SSOToken): Promise<void> {
+        if (!ssoSessionName || !ssoTokenDuckTyper.eval(ssoToken)) {
+            return
+        }
+
+        tryAsync(
+            () => writeSsoObjectToFile(ssoSessionName, ssoToken),
+            error => AwsError.wrap(error, AwsErrorCodes.E_CANNOT_WRITE_SSO_CACHE)
+        )
+    }
+}
+
+async function tryGetAsync<R>(tryIt: () => Promise<R | undefined>): Promise<R | undefined> {
+    try {
+        return await tryIt()
+    } catch (error) {
+        // Error codes are consistent across OSes (Windows is converted to libuv error codes)
+        // https://nodejs.org/api/errors.html#errorerrno
+        if ((error as SystemError)?.code === 'ENOENT') {
+            return Promise.resolve(undefined)
+        }
+
+        throw AwsError.wrap(error as Error, AwsErrorCodes.E_CANNOT_READ_SSO_CACHE)
+    }
+}
+
+function toSsoClientRegistrationCacheId(clientName: string, ssoRegion: string, ssoStartUrl: string): string {
+    return JSON.stringify({
+        region: ssoRegion,
+        startUrl: ssoStartUrl,
+        tool: clientName,
+    })
+}
+
+// Based on:
+// https://github.com/smithy-lang/smithy-typescript/blob/main/packages/shared-ini-file-loader/src/getSSOTokenFromFile.ts
+async function getSsoClientRegistrationFromFile(id: string): Promise<SsoClientRegistration> {
+    const filepath = getSSOTokenFilepath(id)
+    const json = await readFile(filepath, 'utf8')
```

Comment:
No need to catch possible errors here?

Reply:
No, this is just a module private function to help keep things similar to smithy-typescript.  The called getSsoClientRegistration handles any errors thrown.  This function would not be exposed beyond this module.